File: server/pkg/service/publisher-service.go
package service

import (
	"github.com/kirkalyn13/xyz-books-app/pkg/model"
	"github.com/kirkalyn13/xyz-books-app/pkg/repository"
)

// GetPublishers fetches the list of Publishers
func GetPublishers() ([]model.Publisher, error) {
	publishers, err := repository.GetPublishers()

	if err != nil {
		return []model.Publisher{}, err
	}

	return publishers, nil
}

// GetPublishersByID fetches a Publisher based on a given ID
func GetPublisherByID(id string) (model.Publisher, error) {
	publisher, err := repository.GetPublisherByID(id)

	if err != nil {
		return model.Publisher{}, err
	}

	return publisher, nil
}

// AddPublisher adds a new Publisher entity
func AddPublisher(publisher model.Publisher) (model.Publisher, error) {
	result, err := repository.AddPublisher(publisher)

	if err != nil {
		return model.Publisher{}, err
	}

	return result, nil
}

// EditPublisher edits a Publisher entity
func EditPublisher(publisher model.Publisher, id string) (model.Publisher, error) {
	result, err := repository.EditPublisher(publisher, id)

	if err != nil {
		return model.Publisher{}, err
	}

	return result, nil
}

// DeletePublisher deletes a Publisher entity
func DeletePublisher(id string) error {
	err := repository.DeletePublisher(id)

	if err != nil {
		return err
	}

	return nil
}

File: client/src/services/publisherService.ts
import { buildUri } from "../utils/uriBuilder"
import { getEndpoint } from "../utils/getEndpoint"
import { Publisher } from "../types/publisher"
import axios from 'axios'

export const getPublishers = async (q: string): Promise<any> => {
    try {
        const queryParams = {
            q
        }
        return await axios.get(buildUri(getEndpoint("publishers", ""), queryParams))
    } catch(err) {
        return err
    }
}

export const getPublisherByID = async (id: string): Promise<any> => {
    try {
        return await axios.get(buildUri(getEndpoint("publishers", id)))
    } catch(err) {
        return err
    }
}

export const addPublisher = async (publisher: Publisher): Promise<any> => {
    try {
        return await axios.post(buildUri(getEndpoint("publishers", "", "")), publisher)
    } catch(err) {
        return err
    }
}

export const editPublisher = async (id: string, publisher: Publisher): Promise<any> => {
    try {
        return await axios.put(buildUri(getEndpoint("publishers", id, "")), publisher)
    } catch(err) {
        return err
    }
}

export const deletePublisher = async (id: string): Promise<any> => {
    try {
        return await axios.delete(buildUri(getEndpoint("publishers", id)))
    } catch(err) {
        return err
    }
}